core/lib/contract_verifier/src/etherscan/mod.rs

```diff
@@ -0,0 +1,351 @@
+                    if get_status_attempts > MAX_POLL_VERIFICATION_RESULT_ATTEMPTS {
+                        tracing::warn!(
+                            "Get verification status timed out for contract with address {:#?}",
+                            req.contract_address
+                        );
+                        return Err(ProcessingError::VerificationStatusPollingTimeout.into());
+                    }
+                    continue;
+                }
+                Err(e) => return Err(e.into()),
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn reset_etherscan_verification_processing_started_at(
+        &self,
+        request_id: usize,
+    ) -> Result<(), anyhow::Error> {
+        let mut connection = self
+            .connection_pool
+            .connection_tagged("etherscan_verifier")
+            .await?;
+
+        connection
+            .contract_verification_dal()
+            .reset_etherscan_verification_processing_started_at(request_id)
+            .await?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl JobProcessor for EtherscanVerifier {
+    type Job = EtherscanVerificationRequest;
+    type JobId = usize;
+    type JobArtifacts = ();
+    const SERVICE_NAME: &'static str = "etherscan_verifier";
+    const BACKOFF_MULTIPLIER: u64 = 1;
+
+    async fn get_next_job(&self) -> anyhow::Result<Option<(Self::JobId, Self::Job)>> {
+        // We consider that the job is stuck if it's being processed for more than 2 hours.
+        // We need large overhead because Etherscan has pretty strict daily
+        // limits for verification requests.
+        // In case the daily verification rate limit is reached, the processing is paused for 1 hour
+        // and the processing_started_at field is reset to the current time after the pause.
+        // Since etherscan verification is completely async it shouldn't affect user experience much.
+        const TIME_OVERHEAD: Duration = Duration::from_secs(2 * 60 * 60); // 2 hours
+
+        let mut connection = self
+            .connection_pool
+            .connection_tagged(Self::SERVICE_NAME)
+            .await?;
+
+        let job = connection
+            .contract_verification_dal()
+            .get_next_queued_etherscan_verification_request(TIME_OVERHEAD)
+            .await?;
+        Ok(job.map(|job| (job.id, job)))
+    }
+
+    #[allow(clippy::async_yields_async)]
+    async fn process_job(
+        &self,
+        _job_id: &Self::JobId,
+        job: Self::Job,
+        started_at: Instant,
+    ) -> tokio::task::JoinHandle<anyhow::Result<Self::JobArtifacts>> {
+        let this = self.clone();
+        let mut started_at = started_at;
+        tokio::task::spawn(async move {
+            tracing::info!("Started to process request with id = {}", job.id);
+            let mut verify_attempts = 0;
+            let processing_result: Result<(), anyhow::Error> = loop {
```

Comment:
The benefit of `JobProcessor` is that it handles retries for you, but since you have custom logic for retries, maybe your code will become cleaner if you don't use this trait (e.g. you will just write a bunch of methods to pick up / process / save jobs.

Reply:
Agreed, removed `JobProcessor` from the Etherscan verifier.